src/merkle_tree/mod.rs

```diff
@@ -1,209 +1,302 @@
-    Debug(bound = "P: Config, <P::H as FixedLengthCRH>::Output: fmt::Debug")
-)]
+pub type TwoToOneDigest<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Output;
+pub type LeafDigest<P> = <<P as Config>::LeafHash as CRH>::Output;
+pub type TwoToOneParam<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Parameters;
+pub type LeafParam<P> = <<P as Config>::LeafHash as CRH>::Parameters;
+
+/// Stores the hashes of a particular path (in order) from root to leaf.
+/// For example:
+/// ```tree_diagram
+///          A
+///        /  \
+///       B   C
+///     / \  / \
+///    D  E F  H
+///   .. / \ ....
+///     I  J
+/// ```
+///  Suppose we want to prove I, then `leaf_and_sibling_hash` is (I, J), `non_leaf_and_sibling_hash` is
+///  [(B,C),(D,E)]
 pub struct Path<P: Config> {
-    pub(crate) path: Vec<(Digest<P>, Digest<P>)>,
+    pub(crate) leaf_and_sibling_hash: (LeafDigest<P>, LeafDigest<P>),
+    /// The hash path from higher layer to lower layer.
+    pub(crate) non_leaf_and_sibling_hash_path: Vec<(TwoToOneDigest<P>, TwoToOneDigest<P>)>,
+    /// stores the leaf index of the node
+    pub(crate) leaf_index: usize,
 }
 
-pub type Parameters<P> = <<P as Config>::H as FixedLengthCRH>::Parameters;
-pub type Digest<P> = <<P as Config>::H as FixedLengthCRH>::Output;
-
-impl<P: Config> Default for Path<P> {
-    fn default() -> Self {
-        let mut path = Vec::with_capacity(P::HEIGHT as usize);
-        for _i in 1..P::HEIGHT as usize {
-            path.push((
-                <P::H as FixedLengthCRH>::Output::default(),
-                <P::H as FixedLengthCRH>::Output::default(),
-            ));
-        }
-        Self { path }
+impl<P: Config> Path<P> {
+    /// The position of on_path node in `leaf_and_sibling_hash` and `non_leaf_and_sibling_hash_path`.
+    /// `position[i]` is 0 (false) iff `i`th on-path node from top to bottom is on the left.
+    ///
+    /// This function simply converts `self.leaf_index` to boolean array in big endian form.
+    fn position_list(&self) -> Vec<bool> {
+        let position: Vec<_> = (0..self.non_leaf_and_sibling_hash_path.len() + 1)
+            .map(|i| ((self.leaf_index >> i) & 1) != 0)
+            .rev()
+            .collect();
+        position
     }
 }
 
 impl<P: Config> Path<P> {
+    /// Verify that a leaf is at `self.index` of the merkle tree.
+    /// * `leaf_size`: leaf size in number of bytes
     pub fn verify<L: ToBytes>(
         &self,
-        parameters: &<P::H as FixedLengthCRH>::Parameters,
-        root_hash: &<P::H as FixedLengthCRH>::Output,
+        leaf_hash_parameters: &LeafParam<P>,
+        two_to_one_hash_parameters: &TwoToOneParam<P>,
+        root_hash: &TwoToOneDigest<P>,
+        tree_height: usize,
         leaf: &L,
+        leaf_size: usize,
     ) -> Result<bool, crate::Error> {
-        if self.path.len() != (P::HEIGHT - 1) as usize {
+        // verify path length
+        if self.non_leaf_and_sibling_hash_path.len() != tree_height - 2 {
```

Comment:
Should we just infer the tree length from the path length, instead of providing it as an explicit input to the function?

Reply:
Yes. I will fix that

Reply:
fixed